### ISSUE TITLE
Update on Roadmap

### ISSUE THREAD

**Issue by fchollet** [fchollet/keras#426]
Just an update on where Keras is going in the short term.

Here are the goals for the next release:

- More unit test coverage. @phreeza and @tleeuwenburg have been making significant progress on this front.
- Caffe model importing. @pranv has a PR for this, hopefully we can merge it asap.
- Support for automatic architecture/hyperparam search. I am working on it.
- Stateful recurrent layers for efficient sequence generation. See: https://github.com/fchollet/keras/issues/98
- `TimeDistributed` layer to make it possible to include any layer type as part of a recurrent architecture.

Let me know if I'm forgetting anything important.

Regarding visualization tools: these will be developed as a standalone app separate from Keras, in order to be usable with other deep learning frameworks as well. The project is still essentially at the ideation phase. [Hulaos](https://github.com/fchollet/hualos) is a first concrete step.

**Comment by phreeza** [comment]
Probably more long-term, but at one point you mentioned that you see keras going in a direction where it is theano-agnostic. Do you still see it that way? If so, I think some planning in that direction could be helpful,  since it would probably require some kind of middleware, which gets harder to add as the codebase grows.

**Comment by fchollet** [comment]
> Probably more long-term, but at one point you mentioned that you see keras going in a direction where it is theano-agnostic. Do you still see it that way?

Absolutely. This is the only way to ensure the long-term success of the project. Theano will not be forever. The core value that Keras brings (fast prototyping for research purposes) is not Theano-dependent, and so it should be ported to the best platform that will emerge in the future.

As for when we will start working on a switch to an abstract computation backend: as soon as a Theano successor comes along. It might be as soon as this September. We'll see.

> it would probably require some kind of middleware, which gets harder to add as the codebase grows.

Yes. As much as possible the new abstract backend would follow closely the Theano API, so although the conversion would take some effort, it wouldn't be a difficult project. We'll definitely start planning for it in the next couple months.

**Comment by tleeuwenburg** [comment]
Having access to a non-Theano backend would significantly speed up test
execution. Having the unit tests in place will greatly help with verifying
the new backend, also. The tests could run the new backend by default, with
a slow-running test occasionally verifying that all is well on Theano.

On 23 July 2015 at 00:46, François Chollet <notifications@github.com> wrote:

> Probably more long-term, but at one point you mentioned that you see keras
> going in a direction where it is theano-agnostic. Do you still see it that
> way?
>
> Absolutely. This is the only way to ensure the long-term success of the
> project. Theano will not be forever. The core value that Keras brings (fast
> prototyping for research purposes) is not Theano-dependent, and so it
> should be ported to the best platform that will emerge in the future.
>
> As for when we will start working on a switch to an abstract computation
> backend: as soon as a Theano successor comes along. It might be as soon as
> this September. We'll see.
>
> it would probably require some kind of middleware, which gets harder to
> add as the codebase grows.
>
> Yes. As much as possible the new abstract backend would follow closely the
> Theano API, so although the conversion would take some effort, it wouldn't
> be a difficult project. We'll definitely start planning for it in the next
> couple months.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/issues/426#issuecomment-123746342>.
>



-- 
--------------------------------------------------
Tennessee Leeuwenburg
http://myownhat.blogspot.com/
"Don't believe everything you think"

**Comment by futurely** [comment]
What does the "Theano successor" mean exactly? 

NervanaSystems's [backed](https://github.com/NervanaSystems) was the fastest according to [this benchmark](https://github.com/soumith/convnet-benchmarks).

**Comment by fchollet** [comment]
> What does the "Theano successor" mean exactly?

- faster runtime
- no compilation or faster compilation
- supports autodifferentiation

> NervanaSystems's backed was the fastest according to this benchmark.

For convolutions, yes. For now they don't support autodiff, though.

**Comment by harpone** [comment]
@fchollet:
> As for when we will start working on a switch to an abstract computation backend: **as soon as a Theano successor comes along**. It might be as soon as this September. We'll see.

So what exactly *is* the successor? (I'm just struggling with theano.scan and was just sort of wondering...)

**Comment by pranv** [comment]
@harpone, scroll up

**Comment by harpone** [comment]
\*scrolling, scrolling, scrolling...\*

**Comment by pranv** [comment]
>What does the "Theano successor" mean exactly?
* faster runtime
* no compilation or faster compilation
* supports autodifferentiation

**Comment by harpone** [comment]
@pranv yes, I can read. That doesn't answer my question, but nevermind...

**Comment by elanmart** [comment]
http://joschu.github.io/ ?

**Comment by pranv** [comment]
@elanmart, looks great! But no proper GPU support yet. And the code base is very young. But definitely a strong candidate for alternative backend.

**Comment by fchollet** [comment]
@elanmart It looks very nice, we will definitely be keeping tabs on it. Real GPU support is apparently coming soon.

Once GPU support is here and once the library has been a little bit battle-tested, we'll consider whether to port Keras to CGT.

**Comment by futurely** [comment]
Nervana is still the fastest.
https://github.com/soumith/convnet-benchmarks/issues/56
https://github.com/soumith/convnet-benchmarks

**Comment by scott-gray** [comment]
We now have working autodiff in the new neon/nervanagpu code base.  It's not quite as slick yet as CGT but we're taking a close look at that and see no reason why the next version can't be just as functional.

**Comment by pranv** [comment]
Many options for keras backend. Future looking good...!

great for python ecosystem too

**Comment by fchollet** [comment]
@scott-gray Awesome, that's exciting! Do you have some documentation and code examples for the autodiff feature? How would you compare the Neon computation backend to Theano in terms of feature coverage?

I know of 3 potential options for a post-Theano Keras backend, and it looks like we will probably pick one within the next 2-3 months. It's great to see that things are moving fast in the Python-tensor space : )

**Comment by scott-gray** [comment]
I'd hold out another week or so for the fully refactored neon release.  The documentation should be fairly complete.

I think I'm covering most everything that theano can do in my backend.  The new conv kernels use a dimshuffle on the filter to speed things up.  I still need to generalize that code for full transpose(axes=()) support, but that won't be too hard.

I also want to fully support any dimension of tensor in reduction and broadcast operations (ew already works) but that might take me a bit of time to get to.  Right now those things just work with 1 or 2.

At the moment, I'm just fleshing out all my conv kernel tiling sizes so performance will be good for a wide range of params.  Trying to make it as hard as possible for the FFT guys to overtake me.. but I suppose it might be inevitable.  Then again, smaller parallel layer configs (ala googlenet) seem to be the fashion now.  It's hard to see FFT catching up there.  There's not enough dimensionality to hide your latencies and the filters are all small.

In another overlooked area, my gemm kernels are killing cublas on RNN networks. The new neon will have lots of examples of those with benchmarks.  

We're also playing with nvrtc to see if we can't speed up pycuda compilation times some, though that overhead is already pretty minimal.  You generally have a few seconds to wait the very first time you run something, but after that everything runs instantly.  Autodif works at runtime with no overhead (after being cached).

**Comment by pranv** [comment]
@fchollet, what do you think about having a separate `backend` module in keras, from which relavent stuff can be imported elsewhere. That way, multiple backends can be supported.

**Comment by elanmart** [comment]
@fchollet Just out of curiosity, what are those 3 potentail backends (cgt, nervana, ?)?

**Comment by phreeza** [comment]
@elanmart plain old Theano I guess?

**Comment by elanmart** [comment]
@phreeza I wouldn't call Theano a 
>post-Theano Keras backend

;)

**Comment by phreeza** [comment]
Right, missed that :-D

**Comment by fchollet** [comment]
@pranv that's exactly how it will be done (Theano will still be an option for backwards compatibility, so what we are aiming at is the ability to plug different backends). And the `backend` module in Keras will have an API that basically follows Theano's (maybe a tad syntactically friendlier).

**Comment by pranv** [comment]
@fchollet, I was trying to get some of keras to work with CGT. Some places are really simple, replacing `T` by `cgt` suffices. 

However, I found some of the things still missing, for example `clip` and `round `. They might require more effort. 
@joschu 

Regardless, future of keras with such flexibility looks really exciting. It can really become the default DL library, both for masses and organizations.

**Comment by LeavesBreathe** [comment]
From the people I've talked to on reddit and more, they frequently recommend neon as it faster for rnn's including lstms compared to Theano. I think they are coming with an update to support many more features that could handle sequence to sequence learning.

**Comment by scott-gray** [comment]
It's pretty high on my list of priorities to finish an embeddings lookup table kernel for bprop.  This allows you to sum the gradient over unique words.   I should be able to avoid atomics and do it deterministicly.  Hope to have that sometime this week.  I believe that's the only major thing we're missing for RNNs.
